Error when resolving "$ref" with "dot slash" paths

## Describe the bug
When loading a schema via IRL, $ref files cannot be loaded when using the "dot slash" path e.g: `./file.json`

## To Reproduce
Loading the following schema via an IRL
```
{
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "$id": "test.json",
  "title": "test",
  "type": "object",
  "examples": [
    {
      "description": "Foo",
      "uuid": "092BBF46-4069-11E7-9DF9-024BEA7C7CAF"
    }
  ],
  "required": [
    "description",
    "uuid"
  ],
  "properties": {
    "description": {
      "type": "string"
    },
    "uuid": {
      "$ref": "./uuid.json"
    }
  }
}
```

Java Code:
```
JsonSchemaFactory jsonSchemaFactory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V201909);
JsonSchema jsonSchema = jsonSchemaFactory.getSchema(SchemaLocation.of("https://example.com/test.json"));
jsonSchema.validate(" {
      "description": "Bar",
      "uuid": "092BBF46-4069-11E7-9DF9-024BEA7C7CAF"
    }");
```

This will lead to a `FileNotFoundException` because it will try to resolve `https://example.com/./uuid.json`
```
com.networknt.schema.JsonSchemaException: Failed to load json schema fromhttps://example.com/./uuid.json
	at com.networknt.schema.JsonSchemaFactory.getMappedSchema(JsonSchemaFactory.java:521)
	at com.networknt.schema.JsonSchemaFactory.getSchema(JsonSchemaFactory.java:464)
	at com.networknt.schema.RefValidator.lambda$getRefSchema$0(RefValidator.java:64)
	at com.networknt.schema.CachedSupplier.get(CachedSupplier.java:36)
	at com.networknt.schema.JsonSchemaRef.getSchema(JsonSchemaRef.java:32)
	at com.networknt.schema.RefValidator.validate(RefValidator.java:173)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:569)
	at com.networknt.schema.PropertiesValidator.validate(PropertiesValidator.java:85)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:569)
	at com.networknt.schema.RefValidator.validate(RefValidator.java:181)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:569)
	at com.networknt.schema.PropertiesValidator.validate(PropertiesValidator.java:85)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:569)
	at com.networknt.schema.ItemsValidator.doValidate(ItemsValidator.java:152)
	at com.networknt.schema.ItemsValidator.validate(ItemsValidator.java:122)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:569)
	at com.networknt.schema.BaseJsonValidator.validate(BaseJsonValidator.java:287)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:879)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:702)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:682)
	at com.networknt.schema.JsonSchema.validate(JsonSchema.java:629)
        ... 4 more
Caused by: java.io.FileNotFoundException: https://example.com/./uuid.json
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1994)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1599)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:223)
	at com.networknt.schema.resource.UriSchemaLoader.openConnectionCheckRedirects(UriSchemaLoader.java:51)
	at com.networknt.schema.resource.UriSchemaLoader.lambda$getSchema$0(UriSchemaLoader.java:36)
	at com.networknt.schema.JsonSchemaFactory.getMappedSchema(JsonSchemaFactory.java:494)
	... 24 more
```

In versions `< 1.3.0` it worked this way. Changing the schema to `"$ref": "uuid.json"` fixes this issue, but it should also work using dot slash paths.

## Env
 - json-schema-validator version: 1.3.1
 - Java: OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode)
 - OS: macOS

--- a/src/main/java/com/networknt/schema/AbsoluteIri.java
+++ b/src/main/java/com/networknt/schema/AbsoluteIri.java
@@ -106,11 +106,21 @@ public class AbsoluteIri {
                     scheme = scheme + 3;
                 }
                 base = parent(base, scheme);
-                while (iri.startsWith("../")) {
-                    base = parent(base, scheme);
-                    iri = iri.substring(3);
+
+                String[] iriParts = iri.split("/");
+                for (int x = 0; x < iriParts.length; x++) {
+                    if ("..".equals(iriParts[x])) {
+                        base = parent(base, scheme);
+                    } else if (".".equals(iriParts[x])) {
+                        // skip
+                    } else {
+                        base = base + "/" + iriParts[x];
+                    }
                 }
-                return base + "/" + iri;
+                if (iri.endsWith("/")) {
+                    base = base + "/";
+                }
+                return base;
             }
         }
     }

--- a/src/test/java/com/networknt/schema/RefValidatorTest.java
+++ b/src/test/java/com/networknt/schema/RefValidatorTest.java
@@ -1,0 +1,106 @@
+/*
+ * Copyright (c) 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.networknt.schema;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.Collections;
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+import com.networknt.schema.SpecVersion.VersionFlag;
+
+/**
+ * Tests for RefValidator.
+ */
+public class RefValidatorTest {
+    @Test
+    void resolveSamePathDotSlash() {
+        String mainSchema = "{\r\n"
+                + "  \"$id\": \"https://www.example.com/schema/test.json\",\r\n"
+                + "  \"$ref\": \"./integer.json\"\r\n"
+                + "}";
+
+        String otherSchema = "{\r\n"
+                + "  \"type\": \"integer\"\r\n"
+                + "}";
+
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(VersionFlag.V202012,
+                builder -> builder.schemaLoaders(schemaLoaders -> schemaLoaders.schemas(
+                        Collections.singletonMap("https://www.example.com/schema/integer.json", otherSchema))));
+        JsonSchema jsonSchema = factory.getSchema(mainSchema);
+        Set<ValidationMessage> messages = jsonSchema.validate("\"string\"", InputFormat.JSON);
+        assertEquals(1, messages.size());
+    }
+
+    @Test
+    void resolveSamePath() {
+        String mainSchema = "{\r\n"
+                + "  \"$id\": \"https://www.example.com/schema/test.json\",\r\n"
+                + "  \"$ref\": \"integer.json\"\r\n"
+                + "}";
+
+        String otherSchema = "{\r\n"
+                + "  \"type\": \"integer\"\r\n"
+                + "}";
+
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(VersionFlag.V202012,
+                builder -> builder.schemaLoaders(schemaLoaders -> schemaLoaders.schemas(
+                        Collections.singletonMap("https://www.example.com/schema/integer.json", otherSchema))));
+        JsonSchema jsonSchema = factory.getSchema(mainSchema);
+        Set<ValidationMessage> messages = jsonSchema.validate("\"string\"", InputFormat.JSON);
+        assertEquals(1, messages.size());
+    }
+
+    @Test
+    void resolveParent() {
+        String mainSchema = "{\r\n"
+                + "  \"$id\": \"https://www.example.com/schema/test.json\",\r\n"
+                + "  \"$ref\": \"../integer.json\"\r\n"
+                + "}";
+        
+        String otherSchema = "{\r\n"
+                + "  \"type\": \"integer\"\r\n"
+                + "}";
+
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(VersionFlag.V202012,
+                builder -> builder.schemaLoaders(schemaLoaders -> schemaLoaders.schemas(
+                        Collections.singletonMap("https://www.example.com/integer.json", otherSchema))));
+        JsonSchema jsonSchema = factory.getSchema(mainSchema);
+        Set<ValidationMessage> messages = jsonSchema.validate("\"string\"", InputFormat.JSON);
+        assertEquals(1, messages.size());
+    }
+
+    @Test
+    void resolveComplex() {
+        String mainSchema = "{\r\n"
+                + "  \"$id\": \"https://www.example.com/schema/test.json\",\r\n"
+                + "  \"$ref\": \"./hello/././world/../integer.json\"\r\n"
+                + "}";
+        
+        String otherSchema = "{\r\n"
+                + "  \"type\": \"integer\"\r\n"
+                + "}";
+
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(VersionFlag.V202012,
+                builder -> builder.schemaLoaders(schemaLoaders -> schemaLoaders.schemas(
+                        Collections.singletonMap("https://www.example.com/schema/hello/integer.json", otherSchema))));
+        JsonSchema jsonSchema = factory.getSchema(mainSchema);
+        Set<ValidationMessage> messages = jsonSchema.validate("\"string\"", InputFormat.JSON);
+        assertEquals(1, messages.size());
+    }
+}

